src/languages/handlebars.js

```diff
@@ -8,37 +8,163 @@ Category: template
 */
 
 export default function(hljs) {
-  var BUILT_INS = {'builtin-name': 'each in with if else unless bindattr action collection debugger log outlet template unbound view yield lookup'};
+  const BUILT_INS = {
+    'builtin-name': [
+      'action',
+      'bindattr',
+      'collection',
+      'component',
+      'concat',
+      'debugger',
+      'each',
+      'each-in',
+      'get',
+      'hash',
+      'if',
+      'in',
+      'input',
+      'link-to',
+      'loc',
+      'log',
+      'lookup',
+      'mut',
+      'outlet',
+      'partial',
+      'query-params',
+      'render',
+      'template',
+      'textarea',
+      'unbound',
+      'unless',
+      'view',
+      'with',
+      'yield'
+    ].join(" ")
+  };
 
-  var IDENTIFIER_PLAIN_OR_QUOTED = {
-    begin: /".*?"|'.*?'|\[.*?\]|\w+/
+  const LITERALS = {
+    literal: [
+      'true',
+      'false',
+      'undefined',
+      'null'
+    ].join(" ")
   };
 
-  var EXPRESSION_OR_HELPER_CALL = hljs.inherit(IDENTIFIER_PLAIN_OR_QUOTED, {
-    keywords: BUILT_INS,
+  // as defined in https://handlebarsjs.com/guide/expressions.html#literal-segments
+  // this regex matches literal segments like ' abc ' or [ abc ] as well as helpers and paths
+  // like a/b, ./abc/cde, and abc.bcd
+  const IDENFIFIER_REGEX = /(".*?"|'.*?'|\[.*?\]|[^\s!"#%&'()*+,.\/;<=>@\[\\\]^`{|}~]+|\.|\/)+/;
+
+  // identifier followed by a equal-sign (without the equal sign)
+  const HASH_PARAM_REGEX = /(".*?"|'.*?'|\[.*?\]|[^\s!"#%&'()*+,.\/;<=>@\[\\\]^`{|}~]+)(?==)/;
+
+  const HELPER_NAME_OR_PATH_EXPRESSION = {
+    begin: IDENFIFIER_REGEX,
+    lexemes: /[\w.\/]+/
+  };
+
+  const HELPER_PARAMETER = hljs.inherit(HELPER_NAME_OR_PATH_EXPRESSION, {
+    keywords: LITERALS
+  });
+
+  const SUB_EXPRESSION = {
+    illegal: /\}\}/,
+    begin: /\(/,
+    end: /\)/
+    // the "contains" is added below when all necessary sub-modes are defined
+  };
+
+  const HASH = {
+    // fka "attribute-assignment", parameters of the form 'key=value'
+    className: 'attr',
+    illegal: /\}\}/,
```

Comment:
Is a single `}` allowed (perhaps it is)?  If not then perhaps this should be just `\}`?

Reply:
After doing some tests, I'd say the `illegal`-statement does not make any sense, neither at this point nor in the sub-expressions-mode.

I have tried a few strings, but I couldn't find any where the `illegal`-regex really cause a match while parsing.

```handlebars
<p>{{helper }}=x}}</p>

and 

<p>{{helper a}}b=x}}</p>
```
are perfectly valid (although the `}}` is the end of the mustache here and everything after that is xml. It is parsed correctly with and without the `illegal`-regex.

In the sub-expressions only time the illegal `}}` really matches is in a string like 

```handlebars
before {{helper (}} 1 2)}} after
```

The effect is that the `1` is highlighted as `name ` and `2` as number. The same happens if I remove the `illegal`-regex.

I think it is best to remove both statements to avoid confusion.

Reply:
Added a commit to remove the `illegal`-regex statements

Reply:
Illegal is intended to say "this character should NEVER occur in the position in this type of grammar"... and if it is found then it will stop highlighting anything further (or consider the language a non-match for auto-detection purposes). (unless someone uses the ignoreIllegal flag of course)

_Yes, if your illegal is the same as your `end` expression that it will serve no purpose at all since end gets the match first._ Is that what was happening here?

Reply:
## In HASH

I am not sure: `HASH` does not have an `end`-regex, but only a `starts`. As far as I can see, the mode itself is only active for the token matching the `begin`-regex. And that can never be `}}`. Then it switches to different modes. 

## In SUB_EXPRESSION 

I made some tests  in the `developer.html`, with added `illegal: /}}/` in SUB_EXPRESSION.

When I set the language to `handlebars` and use 

```handlebars
some text {{log ( }} ) }} some text 
```

as input, every thing is still colored, even the part `) }}`. An added `console.log` shows me that the illegal-token was found. I would have expected everything to be black...

When I set `language` to `auto`, the language is detected as `AppleScript`. :2nd_place_medal: 

The question here is: Would you expect every snippet that causes a parse error in the Handlebars-parser, to match an `illegal`-regex somewhere? 

That would be a lot of work and we would mean rebuilding the Handlebars-compiler with highlight.js techniques. (It already feels like I have done so for a large part).

I think it is OK if some illegal constructs are still highlighted as good as possible. Maybe someone even wants to highlight a broken template...

I would like to leave it like this.